### PR TITLE
Disable analysis options that are incompatible with input source's output geometry  

### PR DESCRIPTION
**Pull request by viddo**
Waypoint for #6886

screen shot 2016-03-29 at 10 18 14

- [x] get geometry type for a node; async, since it might need to lookup the geometry through the SQL API
- [x] update add-analysis view, show loading until geometry type is resolved for given node model
- [x] update create-analysis-options-models, disable types that doesn't have matching input geometry and change desc
- [x] merge changes to cartodb.js SQL class (to know if describeGeom fails), update shrinkwrap properly etc. :point_right: https://github.com/CartoDB/cartodb.js/pull/1219

@xavijam can you review? 
in particular the async get of the geometry type, if you have any other proposals of a simpler way of handling that I'm all ears.

cc @javierarce @urbanphes

### CODE REVIEW

**Comment by viddo** [comment]
Backend tests fail on the CI for some unknown reason, but there are only changes to the frontend in this PR (which pass) so disregard the "default" check for now.

File: lib/assets/javascripts/cartodb3/data/camshaft-reference.js [hunk 1]
```diff
@@ -69,5 +69,35 @@ module.exports = {
...
+
+    return outputGeometry;
+  },
+
+  getValidInputGeometriesForType: function (type, sourceParamName) {
+    var inputGeometries = [];
+
+    switch (type) {
+      case 'buffer':
+        inputGeometries = ['point', 'polygon'];
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/camshaft-reference.js]: We are not taking into account the possibility of multi-polygon, multi-point and multi-line, that's ok?

Comment (javierarce) [inline lib/assets/javascripts/cartodb3/data/camshaft-reference.js]: I think we don't need that level of specificity right now.

**Comment by xavijam** [comment]
Failures about geocoding quotas and isolines stuff, could you give us some light here @rafatower or @ethervoid?:

```
Failures:
09:12:42 
09:12:42   1) Carto::User behaves like user models User#used_geocoding_quota calculates the used geocoder quota in the current billing cycle
09:12:42      Failure/Error: get_user_by_id(@user1.id).get_new_system_geocoding_calls.should == 210
09:12:42        expected: 210
09:12:42             got: 10.0 (using ==)
09:12:42      Shared Example Group: "user models" called from ./spec/models/carto/user_spec.rb:6
09:12:42      # ./spec/models/user_shared_examples.rb:129:in `block (3 levels) in <top (required)>'
09:12:42 
09:12:42   2) Carto::User behaves like user models User#used_geocoding_quota calculates the used geocoding quota for an organization
09:12:42      Failure/Error: @organization.get_new_system_geocoding_calls.should == 230
09:12:42        expected: 230
09:12:42             got: 100.0 (using ==)
09:12:42      Shared Example Group: "user models" called from ./spec/models/carto/user_spec.rb:6
09:12:42      # ./spec/models/user_shared_examples.rb:155:in `block (3 levels) in <top (required)>'
09:12:42 
09:12:42   3) Carto::User behaves like user models User#used_geocoding_quota calculates the used geocoder quota in the current billing cycle including empty requests
09:12:42      Failure/Error: get_user_by_id(@user1.id).get_new_system_geocoding_calls.should == 220
09:12:42        expected: 220
09:12:42             got: 10.0 (using ==)
09:12:42      Shared Example Group: "user models" called from ./spec/models/carto/user_spec.rb:6
09:12:42      # ./spec/models/user_shared_examples.rb:166:in `block (3 levels) in <top (required)>'
09:12:42 
09:12:42   4) Carto::User behaves like user models User#used_here_isolines_quota calculates the used here isolines quota in the current billing cycle
09:12:42      Failure/Error: @user1.get_here_isolines_calls.should == 110
09:12:42        expected: 110
09:12:42             got: 10.0 (using ==)
09:12:42      Shared Example Group: "user models" called from ./spec/models/carto/user_spec.rb:6
09:12:42      # ./spec/models/user_shared_examples.rb:228:in `block (3 levels) in <top (required)>'
09:12:42 
09:12:42   5) Carto::User behaves like user models User#used_here_isolines_quota calculates the used here isolines quota for an organization
09:12:42      Failure/Error: @organization.get_here_isolines_calls.should == 220
09:12:42        expected: 220
09:12:42             got: 100.0 (using ==)
09:12:42      Shared Example Group: "user models" called from ./spec/models/carto/user_spec.rb:6
09:12:42      # ./spec/models/user_shared_examples.rb:240:in `block (3 levels) in <top (required)>'
09:12:42 
09:12:42   6) Carto::User behaves like user models User#used_here_isolines_quota calculates the used here isolines quota in the current billing cycle including empty requests
09:12:42      Failure/Error: @user1.get_here_isolines_calls.should == 120
09:12:42        expected: 120
09:12:42             got: 10.0 (using ==)
09:12:42      Shared Example Group: "user models" called from ./spec/models/carto/user_spec.rb:6
09:12:42      # ./spec/models/user_shared_examples.rb:250:in `block (3 levels) in <top (required)>'
09:12:42 
09:12:42   7) refactored behaviour behaves like user models User#used_geocoding_quota calculates the used geocoder quota in the current billing cycle
09:12:42      Failure/Error: get_user_by_id(@user1.id).get_new_system_geocoding_calls.should == 210
09:12:42        expected: 210
09:12:42             got: 10.0 (using ==)
09:12:42      Shared Example Group: "user models" called from ./spec/models/user_spec.rb:12
09:12:42      # ./spec/models/user_shared_examples.rb:129:in `block (3 levels) in <top (required)>'
09:12:42 
09:12:42   8) refactored behaviour behaves like user models User#used_geocoding_quota calculates the used geocoding quota for an organization
09:12:42      Failure/Error: @organization.get_new_system_geocoding_calls.should == 230
09:12:42        expected: 230
09:12:42             got: 100.0 (using ==)
09:12:42      Shared Example Group: "user models" called from ./spec/models/user_spec.rb:12
09:12:42      # ./spec/models/user_shared_examples.rb:155:in `block (3 levels) in <top (required)>'
09:12:42 
09:12:42   9) refactored behaviour behaves like user models User#used_geocoding_quota calculates the used geocoder quota in the current billing cycle including empty requests
09:12:42      Failure/Error: get_user_by_id(@user1.id).get_new_system_geocoding_calls.should == 220
09:12:42        expected: 220
09:12:42             got: 10.0 (using ==)
09:12:42      Shared Example Group: "user models" called from ./spec/models/user_spec.rb:12
09:12:42      # ./spec/models/user_shared_examples.rb:166:in `block (3 levels) in <top (required)>'
09:12:42 
09:12:42   10) refactored behaviour behaves like user models User#used_here_isolines_quota calculates the used here isolines quota in the current billing cycle
09:12:42      Failure/Error: @user1.get_here_isolines_calls.should == 110
09:12:42        expected: 110
09:12:42             got: 10.0 (using ==)
09:12:42      Shared Example Group: "user models" called from ./spec/models/user_spec.rb:12
09:12:42      # ./spec/models/user_shared_examples.rb:228:in `block (3 levels) in <top (required)>'
09:12:42 
09:12:42   11) refactored behaviour behaves like user models User#used_here_isolines_quota calculates the used here isolines quota for an organization
09:12:42      Failure/Error: @organization.get_here_isolines_calls.should == 220
09:12:42        expected: 220
09:12:42             got: 100.0 (using ==)
09:12:42      Shared Example Group: "user models" called from ./spec/models/user_spec.rb:12
09:12:42      # ./spec/models/user_shared_examples.rb:240:in `block (3 levels) in <top (required)>'
09:12:42 
09:12:42   12) refactored behaviour behaves like user models User#used_here_isolines_quota calculates the used here isolines quota in the current billing cycle including empty requests
09:12:42      Failure/Error: @user1.get_here_isolines_calls.should == 120
09:12:42        expected: 120
09:12:42             got: 10.0 (using ==)
09:12:42      Shared Example Group: "user models" called from ./spec/models/user_spec.rb:12
09:12:42      # ./spec/models/user_shared_examples.rb:250:in `block (3 levels) in <top (required)>'
09:12:42 
09:12:42 Finished in 8 minutes 31.99 seconds
09:12:42 249 examples, 12 failures, 8 pending
```

File: lib/assets/javascripts/cartodb3/data/analysis-definition-node-model.js [hunk 2]
```diff
@@ -9,7 +10,7 @@ var nodeIds = require('./analysis-definition-node-ids.js');
  */
 module.exports = cdb.core.Model.extend({
 
-  initialize: function () {
+  initialize: function (attrs, opts) {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/analysis-definition-node-model.js]: Do we need that for anything?

File: lib/assets/javascripts/cartodb3/data/camshaft-reference.js (hunk 1 above)
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/camshaft-reference.js]: :+1:

File: lib/assets/javascripts/cartodb3/data/analysis-definition-node-model.js [hunk 3]
```diff
@@ -68,6 +69,26 @@ module.exports = cdb.core.Model.extend({
     };
   },
 
+  /**
+   * @param {Boolean} [retryIfRejected = false] - Set to true to force a new deferred value
+   * @return {Object} A promise, which when resolved contains a {String}, e.g. 'line', 'point', 'polygon'
+   */
+  getOutputGeometryType: function (retryIfRejected) {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/analysis-definition-node-model.js]: I'm not used to using promises and deferrers, but looks like a very good solution. I don't know a pretty good one used in the past, so :+1:.

**Comment by xavijam** [comment]
:es:

File: lib/assets/javascripts/cartodb3/data/analysis-definition-node-model.js (hunk 2 above)
Comment (javierarce) [inline lib/assets/javascripts/cartodb3/data/analysis-definition-node-model.js]: No. It seems we forgot to remove those arguments.

File: lib/assets/javascripts/cartodb3/data/camshaft-reference.js (hunk 1 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb3/data/camshaft-reference.js]: we'll probably need to revisit/change these definitions, once @rochoa is available again, we're using these values for now (matching the value of the `simplified_geometry_type` from a [`sql.describeGeom`](https://github.com/CartoDB/cartodb.js/blob/bd9dabe1b639cb98faddcfd8d9992eaf14d0aa76/src/api/sql.js#L502) result.

File: lib/assets/javascripts/cartodb3/data/analysis-definition-node-model.js (hunk 2 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb3/data/analysis-definition-node-model.js]: No, was accidentally left when moving stuff around, now fixed

**Comment by xavijam** [comment]
@viddo @javierarce, @ethervoid has fixed the problem with the failing tests -> https://github.com/CartoDB/cartodb/pull/6986.

**Comment by viddo** [comment]
Oook, this PR is branched of `analysis` however so the fix is not there yet, I'll update that one after merging this one then

**Pull request merged**